r/R/dplyr-join.R

```diff
@@ -117,10 +117,33 @@ handle_join_by <- function(by, x, y) {
   if (is.null(names(by))) {
     by <- set_names(by)
   }
-  # TODO: nicer messages?
-  stopifnot(
-    all(names(by) %in% names(x)),
-    all(by %in% names(y))
-  )
+
+  missing_x_cols <- setdiff(names(by), names(x))
+  if (length(missing_x_cols) > 0) {
+    message <- paste(
+      "Join",
+      ngettext(length(missing_x_cols), "column", "columns"),
+      "must be present in data."
+    )
+    message_x <- paste(
+      oxford_paste(missing_x_cols, quote_symbol = "`"),
+      "not present in x."
+      )
+    abort(c(message, x = message_x))
+  }
+
+  missing_y_cols <- setdiff(by, names(y))
+  if (length(missing_y_cols) > 0) {
+    message <- paste(
+      "Join",
+      ngettext(length(missing_y_cols), "column", "columns"),
+      "must be present in data."
+    )
+    message_y <- paste(
+      oxford_paste(missing_y_cols, quote_symbol = "`"),
+      "not present in y."
+    )
+    abort(c(message, x = message_y))
+  }
```

Comment:
Would it be possible to accumulate message_x and message_y in parallel and then print them both? This might ultimately simplify this (e.g. the "join column must be present in data" would happen once). But more importantly it would give people the feedback about both x and y at the beginning instead of fixing one and then having another pop up.

Reply:
I had a similar thought and I have a hunch this might be the reason for {dplyr} keeping things simpler / a bit more ambiguous. I will look into it.

Reply:
We might need to simplify the header (no more singular/ plural), but that is a minor thing and sacrifice to make for the addition of a second error bullet point.

Reply:
Done
```
> left_join(
+         arrow_table(example_data),
+         arrow_table(example_data),
+         by = c("made_up_colname1", "made_up_colname2")
+       )
 Error: Join columns must be present in data.
x `made_up_colname1` and `made_up_colname2` not present in x.
x `made_up_colname1` and `made_up_colname2` not present in y.
Run `rlang::last_error()` to see where the error occurred.
```